### PR TITLE
3D Descriptors Dragons

### PR DESCRIPTION
**Pull request by thegodone**
for review
This is the second iteration of #1145

### CODE REVIEW

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+#include <Eigen/SVD>
+#include <iostream>
+#include <Eigen/Core>
+#include <Eigen/QR>
+
+
+
+using namespace Eigen;
+namespace RDKit {
+namespace Descriptors{
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: General comments hat about a dedicated namespace for Descriptors3D?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+namespace {
+
+MolData3Ddescriptors moldata3D;
+
+double* retreiveMat(MatrixXd matrix) {
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: RetrieveMat. Why no matrix2double?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+     double* arrayd = matrix.data();
+     return arrayd;
+  }
+
+double* retreiveVect(VectorXd matrix) {
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: RetrievVect. Why not vector2double?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+     double* arrayd = matrix.data();
+     return arrayd;
+
+  }
+
+VectorXd getEigenVect(std::vector<double> v){
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Convert2vectorXd

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+      double* varray_ptr = &v[0];
+      Map<VectorXd> V(varray_ptr,v.size());
+      return V;
+  }
+
+
+double* AppendDouble(double *w, double* Append, int length, int pos){
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: This is not appending, it is inserting

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    return w;
+}
+
+
+double* GetGeodesicMatrix(double* dist, int lag,int numAtoms){
+    int sizeArray=numAtoms*numAtoms;
+    double *Geodesic = new double[sizeArray];
+    for (int i=0; i<sizeArray;i++) {
+      if (dist[i]==lag) Geodesic[i]=1;
+      else  Geodesic[i]=0;
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Don't know what the convention is, but would like to see braces around the two blocks

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    return Geodesic;
+}
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: This function allocates memory so the caller needs to free it. Is Theresa to ensure that we won't get a memory leak accidentally?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+double* get3DautocorrelationDesc(double* dist3D, double* dist, int numAtoms, const ROMol& mol) {
+
+   Map<MatrixXd> dm(dist3D, numAtoms,numAtoms);
+
+   double *w= new double[80];
+
+   std::vector<double> wp= moldata3D.GetRelativePol(mol);
+
+   VectorXd Wp = getEigenVect(wp);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Is it necessary to create the variable wp? Or could yodo

getEigenVect(moldata3D.GetRelativePol(mol)

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+   VectorXd Wu = getEigenVect(wu);
+
+   std::vector<double>  ws =  moldata3D.GetIState(mol);
+
+   VectorXd Ws = getEigenVect(ws);
+
+   std::vector<double> wr= moldata3D.GetRelativeRcov(mol);
+
+   VectorXd Wr = getEigenVect(wr);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Lots of white space that unnecessirily bloats the code.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+   MatrixXd Bi;
+   MatrixXd tmp;
+   double TDBmat[8][10];
+   double dtmp;
+
+  for (int i=0;i<10;i++){
+    double * Bimat = GetGeodesicMatrix(dist, i+1, numAtoms);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: There is no "delete Bimat". Is there a memory leak?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    Map<MatrixXd> Bi(Bimat, numAtoms,numAtoms);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Is there another use in the code where you don't use the geodesic matrix to create a Map<MatrixXd>. If not why don't do this already in the function?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+  // create the Output vector!
+    w= AppendDouble(w, TDBmat[0], 10, 0);
+    w= AppendDouble(w, TDBmat[1], 10, 10);
+    w= AppendDouble(w, TDBmat[2], 10, 20);
+    w= AppendDouble(w, TDBmat[3], 10, 30);
+    w= AppendDouble(w, TDBmat[4], 10, 40);
+    w= AppendDouble(w, TDBmat[5], 10, 50);
+    w= AppendDouble(w, TDBmat[6], 10, 60);
+    w= AppendDouble(w, TDBmat[7], 10, 70);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: is the assignment to w necessary? This flattens the matrix into an array, doesn't it? Why not use a loop?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+std::vector<double> AUTOCORR3D(const ROMol& mol,int confId){
+  PRECONDITION(mol.getNumConformers()>=1,"molecule has no conformers")
+  int numAtoms = mol.getNumAtoms();
+
+  const Conformer &conf = mol.getConformer(confId);
+
+  double *dist = MolOps::getDistanceMat(mol, false);
+  double *dist3D = MolOps::get3DDistanceMat(mol, confId);
+  double *wpol= Get3Dauto(dist3D, dist, numAtoms, mol);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Where is the memory deleted?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+      std::vector<std::string> AUTOCORRNAMES={"TDB01u","TDB02u","TDB03u","TDB04u","TDB05u","TDB06u","TDB07u","TDB08u","TDB09u","TDB10u","TDB01m","TDB02m","TDB03m","TDB04m","TDB05m","TDB06m","TDB07m","TDB08m","TDB09m","TDB10m","TDB01v","TDB02v","TDB03v","TDB04v","TDB05v","TDB06v","TDB07v","TDB08v","TDB09v","TDB10v","TDB01e","TDB02e","TDB03e","TDB04e","TDB05e","TDB06e","TDB07e","TDB08e","TDB09e","TDB10e","TDB01p","TDB02p","TDB03p","TDB04p","TDB05p","TDB06p","TDB07p","TDB08p","TDB09p","TDB10p","TDB01i","TDB02i","TDB03i","TDB04i","TDB05i","TDB06i","TDB07i","TDB08i","TDB09i","TDB10i","TDB01s","TDB02s","TDB03s","TDB04s","TDB05s","TDB06s","TDB07s","TDB08s","TDB09s","TDB10s","TDB01r","TDB02r","TDB03r","TDB04r","TDB05r","TDB06r","TDB07r","TDB08r","TDB09r","TDB10r"};
+      double *res= new double[80];
+      res= get3DautocorrelationDesc(dist3D,dist, numAtoms, mol);
+      return res;
+  }
+
+} //end of anonymous namespace
+
+
+std::vector<double> AUTOCORR3D(const ROMol& mol,int confId){
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Style. All caps names for function

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+    return w;
+
+}
+
+  double* Get3Dauto(double* dist3D,double* dist, int numAtoms, const ROMol& mol){
+
+      std::vector<std::string> AUTOCORRNAMES={"TDB01u","TDB02u","TDB03u","TDB04u","TDB05u","TDB06u","TDB07u","TDB08u","TDB09u","TDB10u","TDB01m","TDB02m","TDB03m","TDB04m","TDB05m","TDB06m","TDB07m","TDB08m","TDB09m","TDB10m","TDB01v","TDB02v","TDB03v","TDB04v","TDB05v","TDB06v","TDB07v","TDB08v","TDB09v","TDB10v","TDB01e","TDB02e","TDB03e","TDB04e","TDB05e","TDB06e","TDB07e","TDB08e","TDB09e","TDB10e","TDB01p","TDB02p","TDB03p","TDB04p","TDB05p","TDB06p","TDB07p","TDB08p","TDB09p","TDB10p","TDB01i","TDB02i","TDB03i","TDB04i","TDB05i","TDB06i","TDB07i","TDB08i","TDB09i","TDB10i","TDB01s","TDB02s","TDB03s","TDB04s","TDB05s","TDB06s","TDB07s","TDB08s","TDB09s","TDB10s","TDB01r","TDB02r","TDB03r","TDB04r","TDB05r","TDB06r","TDB07r","TDB08r","TDB09r","TDB10r"};
+      double *res= new double[80];
+      res= get3DautocorrelationDesc(dist3D,dist, numAtoms, mol);
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Here double[80] is allocated and immediately thrown away. The res is allocated in the function.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+  const Conformer &conf = mol.getConformer(confId);
+
+  double *dist = MolOps::getDistanceMat(mol, false);
+  double *dist3D = MolOps::get3DDistanceMat(mol, confId);
+  double *wpol= Get3Dauto(dist3D, dist, numAtoms, mol);
+
+  std::vector<double> dataVec;
+  for (int i=0;i<80;i++){
+    dataVec.push_back(wpol[i]);
+  }
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Now you convert it into std::vector<double>. Why not have this from the start?

File: rdkit/Chem/Descriptors.py [hunk 17]
```diff
@@ -21,9 +21,9 @@ def _isCallable(thing):
 
 def _setupDescriptors(namespace):
   global _descList, descList
-  from rdkit.Chem import GraphDescriptors, MolSurf, Lipinski, Fragments, Crippen
+  from rdkit.Chem import GraphDescriptors, MolSurf, Lipinski, Fragments, Crippen, Descriptors3D
   from rdkit.Chem.EState import EState_VSA
-  mods = [GraphDescriptors, MolSurf, EState_VSA, Lipinski, Crippen, Fragments]
+  mods = [GraphDescriptors, MolSurf, EState_VSA, Lipinski, Crippen,Descriptors3D, Fragments]
```
Comment (gedeck) [inline rdkit/Chem/Descriptors.py]: Please remove Descriptors3D from this file. We just did this in master.

File: rdkit/Chem/UnitTestDescriptors.py [hunk 18]
```diff
@@ -50,6 +50,9 @@ def testBadAtomHandling(self):
       for nm, fn in Descriptors._descList:
         try:
           v = fn(m)
+        except RuntimeError:
+          # 3D descriptors fail since the mol has no conformers
+          pass
```
Comment (gedeck) [inline rdkit/Chem/UnitTestDescriptors.py]: Same here. This was included to quiet the test for the 3D descriptors. As we removed them, this exception handling is no longer required

File: Code/GraphMol/Descriptors/Data3Ddescriptors.h [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+		static double rcov[83];
+
+   public:
+	   Data3Ddescriptors();
+	   double* getMW();
+	   double* getVDW();
+	   double* getNEG();
+	   double* getPOL();
+	   double* getIonPOL();
+	   double* getRCOV();
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/Data3Ddescriptors.h]: I think some comments are required here. This is not MW, so  a different name or a comment would be useful. Also, where do the values come from.

File: Code/GraphMol/Descriptors/MolData3Ddescriptors.cpp [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+
+MolData3Ddescriptors::MolData3Ddescriptors(){}
+
+
+std::vector<double> MolData3Ddescriptors::GetUn(int numAtoms){
+
+  std::vector<double> u(numAtoms, 1.0);
+
+  return u;
+}
```
Comment (gedeck) [inline Code/GraphMol/Descriptors/MolData3Ddescriptors.cpp]: What is the rationale to create such a simple function?

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 21]
```diff
@@ -0,0 +1,240 @@
+//
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: This copyright statement is wrong.
Please include the correct year and your own name (if you can do that) or nothing at all and I'll add something later.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp (hunk 7 above)
Comment (greglandrum) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: It does place a burden on the caller, so it's dangerous. 
I will run the code inside valgrind to detect leaks.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+
+double* get3DautocorrelationDesc(double* dist3D, double* dist, int numAtoms, const ROMol& mol) {
+
+   Map<MatrixXd> dm(dist3D, numAtoms,numAtoms);
+
+   double *w= new double[80];
+
+   std::vector<double> wp= moldata3D.GetRelativePol(mol);
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: all of these calls are making copies of std::vectors. This is certainly extra work and memory allocation that's not needed.
It's better to use references or const references when you can.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp (hunk 9 above)
Comment (greglandrum) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: clang-format will resolve that automatically

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+
+  for (int i=0;i<10;i++){
+    double * Bimat = GetGeodesicMatrix(dist, i+1, numAtoms);
+    Map<MatrixXd> Bi(Bimat, numAtoms,numAtoms);
+    MatrixXd RBi=Bi.cwiseProduct(dm);
+    double Bicount=(double) Bi.sum();
+
+    tmp=Wu.transpose() * RBi * Wu / Bicount;
+    dtmp=(double)tmp(0);
+    if  (std::isnan(dtmp)) dtmp=0.0;
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: I'm not familiar enough with the descriptor to be sure, but this is not always a great solution for dealing with NaN. Are you sure that this is right?

File: Code/GraphMol/Descriptors/Data3Ddescriptors.cpp [hunk 24]
```diff
@@ -0,0 +1,42 @@
+#include <cstdlib>
+#include <iostream>
+#include "Data3Ddescriptors.h"
+
+using namespace std;
+
+
+double Data3Ddescriptors::mw[53]={0.084,0,0,0,0.900,1.000,1.166,1.332,1.582,0,0,0, 2.246, 2.339, 2.579,2.670, 2.952,0,0,0,0,0,0,0,0, 4.650, 4.907, 4.887, 5.291, 5.445,0,0,0,0, 6.653,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 9.884,0,0, 10.56};
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/Data3Ddescriptors.cpp]: A general comment on these arrays: they are all defined to be a fixed size, but in the rest of the code I don't think you actually check to make sure that the indices into them are valid. So, for example, if I have a molecule with Th in it, I will end up with a random value if I ask for the vdW value of it.
That bounds checking is needed somewhere...

File: rdkit/Chem/UnitTestDescriptors.py (hunk 18 above)
Comment (thegodone) [inline rdkit/Chem/UnitTestDescriptors.py]: I need to make the test if I want to correct my code. Want should I do to be able to compute them.

File: rdkit/Chem/Descriptors.py (hunk 17 above)
Comment (thegodone) [inline rdkit/Chem/Descriptors.py]: see next comment on UnitTest

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp (hunk 7 above)
Comment (bp-kelley) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Returning a vector of doubles from a function is free in c++0x and c++11 (there is no copy), I highly recommend this unless there is a really good reason not to do it.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp (hunk 22 above)
Comment (bp-kelley) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: if moldata3D.getRelativePol is returning a vector, surprisingly there is no copy here.  In C++0x The compiler is smart enough to know that the vector on the function stack should be the same as the target vector outside the function and magically makes it work (look up copy elision).  In C++11 this uses the move constructor, so basically, I would say don't worry about it, it makes for much saner APIs.

Comment (bp-kelley) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: Note that this is only for the form

```
vector<double> foo() {...};

vector<double> bar = foo();
```

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+    w= AppendDouble(w, TDBmat[6], 10, 60);
+    w= AppendDouble(w, TDBmat[7], 10, 70);
+
+    return w;
+
+}
+
+  double* Get3Dauto(double* dist3D,double* dist, int numAtoms, const ROMol& mol){
+
+      std::vector<std::string> AUTOCORRNAMES={"TDB01u","TDB02u","TDB03u","TDB04u","TDB05u","TDB06u","TDB07u","TDB08u","TDB09u","TDB10u","TDB01m","TDB02m","TDB03m","TDB04m","TDB05m","TDB06m","TDB07m","TDB08m","TDB09m","TDB10m","TDB01v","TDB02v","TDB03v","TDB04v","TDB05v","TDB06v","TDB07v","TDB08v","TDB09v","TDB10v","TDB01e","TDB02e","TDB03e","TDB04e","TDB05e","TDB06e","TDB07e","TDB08e","TDB09e","TDB10e","TDB01p","TDB02p","TDB03p","TDB04p","TDB05p","TDB06p","TDB07p","TDB08p","TDB09p","TDB10p","TDB01i","TDB02i","TDB03i","TDB04i","TDB05i","TDB06i","TDB07i","TDB08i","TDB09i","TDB10i","TDB01s","TDB02s","TDB03s","TDB04s","TDB05s","TDB06s","TDB07s","TDB08s","TDB09s","TDB10s","TDB01r","TDB02r","TDB03r","TDB04r","TDB05r","TDB06r","TDB07r","TDB08r","TDB09r","TDB10r"};
```
Comment (bp-kelley) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]: This won't compile on non-c++11 compilers.

File: Code/GraphMol/Descriptors/AUTOCORR3D.cpp (hunk 13 above)
Comment (bp-kelley) [inline Code/GraphMol/Descriptors/AUTOCORR3D.cpp]:
```
  rdk_auto_ptr<double> dist = MolOps::getDistanceMat(mol, false);
```

makes this pattern easier and cleans up after itself.

File: Code/GraphMol/Descriptors/Data3Ddescriptors.cpp [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+//double Data3Ddescriptors::vdW[53]={0.299,0,0,0,0.796,1.000,0.695,0.512,0.410,0,0,0, 1.626, 1.424, 1.181,1.088, 1.035,0,0,0,0,0,0,0,0, 1.829, 1.561, 0.764, 0.512, 1.708,0,0,0,0, 1.384,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 2.042,0,0, 1.728}; // official book version
+double Data3Ddescriptors::vdW[83]={0.263,0,1.227,0,0.869,1,0.758,0.715,0.647,0,2.381,1.054,1.774,1.885,1.289,1.187,1.091,0,4.233,0,0,0,0,2.167,2.109,1.995,1.703,0.834,0.558,0.547,1.331,1.396,1.289,1.396,1.289,0,0,0,0,0,0,1.628,0,0,0,0,1.035,0.803,1.463,2.227,1.885,1.779,1.580,0,0,0,0,0,0,0,0,0,0,3.536,0,0,0,0,0,0,0,0,0,0,0,0,0,1.091,0.931,0.758,1.533,1.678,1.885}; // try correction of values
```
Comment (bp-kelley) [inline Code/GraphMol/Descriptors/Data3Ddescriptors.cpp]: these should probably be const.  Again, they are c++11 only initializers.

**Comment by thegodone** [comment]
Dear Greg,

I try to correct AUTOCORR3D, RDF, MORSE and simplify GETAWAY & WHIM computations.

- AUTOCORR is not equal to Dragon (even the article is the same)
- RDF & MORSE are equal to Dragon except for the weigthed States (Istate are not the same in Dragon & RDkit?)
- GETAWAY is equal or close to Dragon except for the weigthed States, I fix RCON computation, I have issues with clustering ITH, ISH values in some cases no sure why.
- WHIM is close to Dragon  except for the weigthed States (again).

GETAWAY and WHIM not converted to void but still std:vector of double

Tell me If we can do better, 

BR

Guillaume

File: .gitignore [hunk 27]
```diff
@@ -71,6 +71,7 @@ __pycache__/
 /Code/GraphMol/test_data/prop.pkl
 
 /Projects/DbCLI/testData/bzr/
-
+External/INCHI-API/*
```
Comment (greglandrum) [inline .gitignore]: This directory is part of the RDKit source. It cannot be in .gitgnore

File: Code/GraphMol/Descriptors/testAUTOCORR2D.cpp [hunk 28]
```diff
@@ -0,0 +1,119 @@
...
+
+  std::string line;
+  std::vector<std::vector<std::string>> data;
+
+  while(std::getline(instrm, line)) {
+      std::string phrase;
+      std::vector<std::string> row;
+      std::stringstream ss(line);
+      while(std::getline(ss, phrase, '\t')) {
+          row.push_back(std::move(phrase));
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/testAUTOCORR2D.cpp]: std::move is a C++11 feature.

File: Code/GraphMol/Descriptors/testGETAWAY.cpp [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+
+void testGETAWAY(int precision) {
+  std::cout << "=>start test GETAWAY\n";
+
+  std::string pathName = getenv("RDBASE");
+  std::string sdfName =
+      pathName + "/Code/GraphMol/Descriptors/test_data/PBF_egfr.sdf";
+
+  RDKit::SDMolSupplier reader(sdfName, true, false);
+  std::string fName = pathName+"/Code/GraphMol/Descriptors/test_data/GETAWAY.out";
```
Comment (greglandrum) [inline Code/GraphMol/Descriptors/testGETAWAY.cpp]: should be `getaway.out`

**Comment by greglandrum** [comment]
In order to detect memory problems, I've run valgrind across:
  - testRDF: no problems
  - testMORSE: no problems
  - testWHIM: problems
  - testGETAWAY: major memory leak
  - testAUTOCORR3D: major memory leak
  - testAUTOCORR2D: memory corruption

more details coming

**Comment by greglandrum** [comment]
Here's the type of error message (there are a bunch of them) that show up for testWHIM:
```
==21982== Invalid read of size 8
==21982==    at 0x4E9D928: RDKit::Descriptors::(anonymous namespace)::getWhimD(std::vector<double, std::allocator<double> >, Eigen::Matrix<double, -1, -1, 0, -1, -1>, int, double) [clone .constprop.546] (WHIM.cpp:147)
==21982==    by 0x4E9FB59: GetWHIMs (WHIM.cpp:279)
==21982==    by 0x4E9FB59: getWHIM (WHIM.cpp:332)
==21982==    by 0x4E9FB59: RDKit::Descriptors::WHIM(RDKit::ROMol const&, std::vector<double, std::allocator<double> >&, int, double) (WHIM.cpp:365)
==21982==    by 0x406365: testWHIM() (testWHIM.cpp:155)
==21982==    by 0x40324D: main (testWHIM.cpp:194)
==21982==  Address 0x87e0c10 is 0 bytes inside a block of size 24 free'd
==21982==    at 0x4C2EDEB: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==21982==    by 0x4E9D7B2: aligned_free (Memory.h:243)
==21982==    by 0x4E9D7B2: conditional_aligned_free<true> (Memory.h:317)
==21982==    by 0x4E9D7B2: conditional_aligned_delete_auto<double, true> (Memory.h:445)
==21982==    by 0x4E9D7B2: ~DenseStorage (DenseStorage.h:403)
==21982==    by 0x4E9D7B2: ~PlainObjectBase (PlainObjectBase.h:88)
==21982==    by 0x4E9D7B2: ~Matrix (Matrix.h:127)
==21982==    by 0x4E9D7B2: RDKit::Descriptors::(anonymous namespace)::getWhimD(std::vector<double, std::allocator<double> >, Eigen::Matrix<double, -1, -1, 0, -1, -1>, int, double) [clone .constprop.546] (WHIM.cpp:143)
==21982==    by 0x4E9FB59: GetWHIMs (WHIM.cpp:279)
==21982==    by 0x4E9FB59: getWHIM (WHIM.cpp:332)
==21982==    by 0x4E9FB59: RDKit::Descriptors::WHIM(RDKit::ROMol const&, std::vector<double, std::allocator<double> >&, int, double) (WHIM.cpp:365)
==21982==    by 0x406365: testWHIM() (testWHIM.cpp:155)
==21982==    by 0x40324D: main (testWHIM.cpp:194)
==21982==  Block was alloc'd at
==21982==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==21982==    by 0x4E9D740: aligned_malloc (Memory.h:220)
==21982==    by 0x4E9D740: conditional_aligned_malloc<true> (Memory.h:301)
==21982==    by 0x4E9D740: conditional_aligned_new_auto<long int, true> (Memory.h:423)
==21982==    by 0x4E9D740: DenseStorage (DenseStorage.h:386)
==21982==    by 0x4E9D740: PlainObjectBase (PlainObjectBase.h:471)
==21982==    by 0x4E9D740: Matrix (Matrix.h:300)
==21982==    by 0x4E9D740: RDKit::Descriptors::(anonymous namespace)::getWhimD(std::vector<double, std::allocator<double> >, Eigen::Matrix<double, -1, -1, 0, -1, -1>, int, double) [clone .constprop.546] (WHIM.cpp:143)
==21982==    by 0x4E9FB59: GetWHIMs (WHIM.cpp:279)
==21982==    by 0x4E9FB59: getWHIM (WHIM.cpp:332)
==21982==    by 0x4E9FB59: RDKit::Descriptors::WHIM(RDKit::ROMol const&, std::vector<double, std::allocator<double> >&, int, double) (WHIM.cpp:365)
==21982==    by 0x406365: testWHIM() (testWHIM.cpp:155)
==21982==    by 0x40324D: main (testWHIM.cpp:194)
```

File: Code/GraphMol/Descriptors/testAUTOCORR2D.cpp (hunk 28 above)
Comment (thegodone) [inline Code/GraphMol/Descriptors/testAUTOCORR2D.cpp]: This is not ready to be implemented there is an issue during the computation, I don't know where

File: .gitignore (hunk 27 above)
Comment (thegodone) [inline .gitignore]: ok

File: Code/GraphMol/Descriptors/testGETAWAY.cpp (hunk 29 above)
Comment (thegodone) [inline Code/GraphMol/Descriptors/testGETAWAY.cpp]: ok

**Comment by thegodone** [comment]
How to solve the issues on WHIM ? 

How can I compile with valgrind to check by myself?

**Comment by greglandrum** [comment]
The error messages indicate that memory that has already been freed is being used. I would guess that there's something wrong with the way the eigen code is being used.

To use valgrind: you just need to compile the code with debug symbols turned on. You can do this by telling cmake to build the code in Debug mode (instead of Release mode, which is the default). After rebuilding you'll have the symbols and you can do: `valgrind testWHIM`

File: .gitignore (hunk 27 above)
Comment (gedeck) [inline .gitignore]: I use `~/.gitignore_global` to keep directories like this or files that are created by the IDE from appearing in a list of potential files to be added to github. For example, I have these two lines in it.
```
/External/AvalonTools/*
/External/INCHI-API/*
```

**Comment by thegodone** [comment]
No luck valgrid is not compatible with sierra

Envoyé de mon iPhone

**Comment by thegodone** [comment]
Ww can close this pull request.

**Pull request closed without merging**